### DISCUSSION TITLE
Where are testcases that timeout stored?

### DISCUSSION THREAD

**Discussion by MathisMARION** [AFLplusplus/AFLplusplus#1378] [Q&A]
Hello and thank you to the community maintaining this great project!
I have a short question.

The status screen displays how many timeouts have occurred during fuzzing.
I am seeing a surprisingly high number of timeouts (about 1% of cases), and I would like to investigate.
The screens says that some testcases have been saved, but I could not find where...
```
total tmouts : 153k (67 saved)
```

**Comment by domenukk** [comment]
They should be in the `hangs` folder of the output directory

**Reply by MathisMARION** [reply]
That is what I thought, but the folder is empty...
AFL displays that it has found no hangs, just timeouts by the way.
```
saved hangs : 0
```

**Reply by vanhauser-thc** [reply]
if a timeout is found it is rerun and given more time to see if it is real hang or just takes longer than allowed. hangs are recorded, longer running inputs not.

**Reply by MathisMARION** [reply]
I see, I got confused by the status screen saying timeouts were saved.
If I understood they are only saved internally to identify potential hangs, which will then be verified.

I still find it surprising that I get so many timeouts though, do you have any suggestion for me to investigate the issue?
I was thinking maybe some testcases from the queue could be identified...

**Reply by vanhauser-thc** [reply]
only hangs are saved, at "timeouts" in the UI it nowhere says that these are saved too :)
you could increase the timeout (-t parameter). and after some time check with `time ./target out/default/queue/id:000....` what the runtime is and isolate those that take long

**Reply by MathisMARION** [reply]
Well in my UI, I do see one line `total tmouts`which says that some cases are being saved:
```
           american fuzzy lop ++4.01a {default} (build/wsbrd) [fast]
┌─ process timing ────────────────────────────────────┬─ overall results ────┐
│        run time : 0 days, 23 hrs, 54 min, 18 sec    │  cycles done : 20    │
│   last new find : 0 days, 0 hrs, 1 min, 38 sec      │ corpus count : 720   │
│last saved crash : none seen yet                     │saved crashes : 0     │
│ last saved hang : none seen yet                     │  saved hangs : 0     │
├─ cycle progress ─────────────────────┬─ map coverage┴──────────────────────┤
│  now processing : 11.107 (1.5%)      │    map density : 6.07% / 12.31%     │
│  runs timed out : 0 (0.00%)          │ count coverage : 3.24 bits/tuple    │
├─ stage progress ─────────────────────┼─ findings in depth ─────────────────┤
│  now trying : splice 3               │ favored items : 66 (9.17%)          │
│ stage execs : 98/110 (89.09%)        │  new edges on : 83 (11.53%)         │
│ total execs : 73.2M                  │ total crashes : 0 (0 saved)         │
│  exec speed : 1025/sec               │  total tmouts : 648k (71 saved)     │
├─ fuzzing strategy yields ────────────┴─────────────┬─ item geometry ───────┤
│   bit flips : disabled (default, enable with -D)   │    levels : 18        │
│  byte flips : disabled (default, enable with -D)   │   pending : 117       │
│ arithmetics : disabled (default, enable with -D)   │  pend fav : 0         │
│  known ints : disabled (default, enable with -D)   │ own finds : 719       │
│  dictionary : n/a                                  │  imported : 0         │
│havoc/splice : 313/26.0M, 406/47.0M                 │ stability : 97.41%    │
│py/custom/rq : unused, unused, unused, unused       ├───────────────────────┘
│    trim/eff : 13.76%/228k, disabled                │          [cpu000: 33%]
└────────────────────────────────────────────────────┘
```

I guess I could write a script going through all testcases to see which ones take longer, but I expected AFL to already provide an indication on slow testcases given the information displayed.

**Reply by vanhauser-thc** [reply]
weird you are right. I have to check this.

**Reply by vanhauser-thc** [reply]
I found something weird. "saved" is wrong, they are actually thrown away even if they have new coverage. so I will be running a test on fuzzbench if the fuzzer improves even they are really saved. either way this will be fixed by either really saving them, or removing the "saved" displayed.

**Reply by MathisMARION** [reply]
Interesting... I'm glad I asked my question then.
Thank you for investigating!

**Reply by vanhauser-thc** [reply]
results are in: https://www.fuzzbench.com/reports/experimental/2022-04-07-aflpp/index.html (llvm15: keep timeouts, other: throw away)
it is a minimal difference. keeping timeout (which is currently not happening) is most of the time slightly better, however for sqllite it makes it so worse that the overall stats are negative in comparison.
I think I will make this an option to save them (but it will not be the default).

**Reply by vanhauser-thc** [reply]
dev branch now has a AFL_KEEP_TIMEOUT option. and without activating this `(0 saved)` will be seen in the timeout counting. thanks for pointing this out
